Replace KJ_STACK_ARRAY with kj::SboArray and kj::SboArrayBuilder

harrishancock opened the pull request:
We'd like to be able to use KJ_STACK_ARRAY in coroutines, but KJ_STACK_ARRAY uses variable-length arrays on compilers which support them (that is: GCC). ~~So, we're dropping support for VLAs from KJ_STACK_ARRAY.~~

It's worth noting that at present, we don't actually support GCC, so we have no target which even uses VLAs anymore. But, we'd like to support GCC again in the future, once its coroutine support is fixed.

Update: Instead of dropping support for VLAs from KJ_STACK_ARRAY, we're replacing the macro entirely with `kj::SboArray`.

harrishancock commented:
Bazel builds broke because a new version of zlib came out.

It's kind of nice that we're notified of new versions of zlib due to breaking builds, but also kind of not, so I updated the link to a stable URL.

kentonv commented:
As discussed, at this point we could make this a template type that wouldn't have to be stack-specific (indeed, that is the whole motivation -- coroutines aren't stack frames). Maybe `SmallArray<T, n>`?

harrishancock commented:
> As discussed, at this point we could make this a template type that wouldn't have to be stack-specific (indeed, that is the whole motivation -- coroutines aren't stack frames). Maybe SmallArray<T, n>?

I've updated this PR with an implementation of what I called `SboArray<T, n>`, and a related `SboArrayBuilder<T, n>`. (I'm happy to switch to SmallArray, I forgot about that suggestiohn.)

I implemented the new types in terms of a new `_::SboArrayDisposer` which is similar in shape, and shares some implementation with, `HeapArrayDisposer`. I did this because I wanted to make their behaviors as similar as possible to `heapArray()` and `heapArrayDisposer()`, which are the implementations we fall back to if the maximum inline size is exceeded. This may have ended up being overkill, though.

While testing, I noticed that `~Array()` is `noexcept`, while `~ArrayBuilder()` is `noexcept(false)`. Should they both instead be `      noexcept(noexcept(instance<T&>().~T()))`?

kentonv commented:
> I noticed that ~Array() is noexcept

Wha... that seems like a huge mistake, surprised I never noticed.

kentonv commented on c++/src/kj/array.h:
```diff
@@ -573,6 +601,81 @@ class CappedArray {
+
+
+template <typename T, size_t smallSize>
+class SboArray: private _::SboArrayDisposer<T, smallSize>,
```
Wow... this is... quite an implementation. I would not have thought of this, but it turns out pretty concise.

I guess inheriting from SboArrayDisposer forces the SboArray to contain an extra pointer for the virtual table for this disposer. I think if you placed `T space[smallSize];` inside `SboArray` itself, then you could reference a singleton disposer, eliminating this pointer?

harrishancock replied:
Yeah, that's a better idea. I pushed 5e8076992236819efdd9aec1e6094cf619853803 with that change, plus a few more cleanup commits to simplify the allocation logic.

harrishancock commented:
Force-pushes were mostly rebases on `v2` and renaming `SboArray` to `SmallArray`. Additionally, I reverted my attempted modernization of the disposal/allocation implementation details -- my simplification caused `heapArray<const byte>()` to become a compile error in workerd, somehow.
